src/status_im/utils/mixpanel.cljs

```diff
@@ -1,35 +1,73 @@
             [goog.crypt.base64 :as b64]
             [status-im.utils.build :as build]
             [status-im.utils.config :as config]
             [status-im.utils.datetime :as datetime]
             [status-im.utils.http :as http]
             [status-im.utils.platform :as platform]
-            [status-im.utils.types :as types]))
+            [status-im.utils.types :as types]
+            [status-im.react-native.js-dependencies :as rn-dependencies]))
 
 (def base-url "http://api.mixpanel.com/")
 (def base-track-url (str base-url "track/"))
 
 (defn encode [m]
   (b64/encodeString (types/clj->json m)))
 
-(defn- build-url [id label props]
-  (str base-track-url
-       "?data="
-       (encode {:event      label
-                :properties (merge
-                              {:token       config/mixpanel-token
-                               :distinct_id id
-                               :os          platform/os
-                               :os-version  platform/version
-                               :app-version build/version
-                               :time        (datetime/timestamp)}
-                              props)})))
-
-(defn track [id label props]
-  (http/get (build-url id label props) nil identity))
+(defn- make-event [id label props]
+  {:event      label
+   :properties (merge
+                {:token       config/mixpanel-token
+                 :distinct_id id
+                 :os          platform/os
+                 :os-version  platform/version
+                 :app-version build/version
+                 :time        (datetime/timestamp)}
+                props)})
+
+(def ^:private pending-events-queue (async/chan (async/sliding-buffer 2000)))
+
+(defn- submit-batch [events]
+  (let [done-chan (async/chan)]
+    (-> (rn-dependencies/fetch
+         base-track-url
+         (clj->js {:method "POST"
+                   :body (str "data=" (encode (into [] events)))
+                   :timeout http/http-request-default-timeout-ms}))
+        (.then  (fn [_] (async/go (async/close! done-chan))))
+        (.catch (fn [_] (async/go (async/close! done-chan)))))
+    done-chan))
+
+(def max-batch-size 50)
+
+(defn- drain-events-queue! []
+  (let [done-chan (async/chan)]
+    (async/go-loop [accumulator []]
+      (let [[event _] (async/alts! [pending-events-queue] :default ::drained)]
+        (cond
+          (= event ::drained)
+          (do (when (not= [] accumulator)
+                (async/<! (submit-batch accumulator)))
+              (async/close! done-chan))
+
+          (= (- max-batch-size 1) (count accumulator))
+          (do (async/<! (submit-batch (conj accumulator event)))
+              (recur []))
+
+          :else
+          (recur (conj accumulator event)))))
+    done-chan))
+
+(defn track [id label props offline?]
+  (let [event (make-event id label props)]
+    (if offline?
+      (async/go (async/>! pending-events-queue event))
+      (async/go
+        (async/<! (drain-events-queue!))
```

Comment:
Then the channel will be immediately closed and nothing happens.

Reply:
To clarify: we try and drain on the next tracking event that happens while online.

Reply:
I guess that's good enough. Can you add a comment here too? :)

Reply:
Sure!

Reply:
Hi @msuess, thanks for the PR, looking really good!

I was wondering maybe there are a few things we can simplify:

We probably don't need `done-chan`, as `go-loop` will return a chan which will be populated with the return value of the loop, and closed after:

`(a/<!! (go-loop [] (Thread/sleep 2000))` waits for 2 seconds

You can also use `poll!` instead of `alts!` and a default value (poll! will return nil if the channel is empty)

We also tend to use `seq` to check if a collection is not empty

something like:

```
(defn- drain-events-queue! []
  (async/go-loop [accumulator []]
    (let [event (async/poll! pending-events-queue)]
      (cond
        (nil? event)
        (when (seq accumulator)
              (async/<! (submit-batch accumulator)))

        (= (- max-batch-size 1) (count accumulator))
        (do (async/<! (submit-batch (conj accumulator event)))
            (recur []))
        :else
        (recur (conj accumulator event))))))
```

I would personally also use `partition-all`, which makes the code a bit more compact (there some cpu tradeoff, but don't think has an impact):
```
(defn- drain-events-queue! []
  (let [events (loop [accumulator []]
                 (if-let [event (async/poll! pending-events-queue)]
                   (recur (conj accumulator event))
                   accumulator))]
    (async/go
              (doseq [batch (partition-all max-batch-size events)] 
                (async/<! (submit-batch batch))))))
```

`partion-all` can also be applied as a transducer to the channel `(chan 100 (partition-all 2))`, which means we would not have to loop and recur as elements would be batched, but unfortunately we would need to close the channel to process everything, I can't think of a better way, but if you can think of a neat solution go for it!

Not sure about this one, but we could also always push to the queue, and drain when we are online:

```
  (async/go (async/>! pending-events-queue event))
  (when-not offline?
   (async/go
    (async/<! (drain-events-queue!)) ; without the http/get

```

what do you think?